pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+	}
+	c.logger.Log("info", "finished syncing git mirror")
+}
+
+// processChangedMirror syncs all given `v1.HelmRelease`s with the
+// mirror we received a change signal for and schedules a release,
+// but only if the sync indicated the change was relevant.
+func (c *GitChartSync) processChangedMirror(mirror string, repo *git.Repo, hrs []*v1.HelmRelease) {
+	for _, hr := range hrs {
+		if _, ok, _ := c.sync(hr, mirror, repo); ok {
+			cacheKey, err := cache.MetaNamespaceKeyFunc(hr.GetObjectMeta())
+			if err != nil {
+				continue // this should never happen
+			}
+			// Schedule release sync by adding it to the queue.
+			c.releaseQueue.AddRateLimited(cacheKey)
+		}
+	}
+}
+
+func (c *GitChartSync) get(hr *v1.HelmRelease) (sourceRef, bool) {
+	c.sourcesMu.RLock()
+	defer c.sourcesMu.RUnlock()
+	if s, ok := c.sources[hr.ResourceID().String()]; ok && s.forHelmRelease(hr) {
+		return s, ok
+	}
+	return sourceRef{}, false
+}
+
+func (c *GitChartSync) store(hr *v1.HelmRelease, s sourceRef) {
+	c.sourcesMu.Lock()
+	c.sources[hr.ResourceID().String()] = s
+	c.sourcesMu.Unlock()
+}
+
+// sync synchronizes the record we have for the given `v1.HelmRelease`
+// with the given mirror. It always updates the HEAD record in the
+// `sourceRef`, but only returns `true` if the update was relevant for
+// the release (e.g. a change in git the chart source path, or a new
+// record). In case of failure it returns an error.
+func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Repo) (sourceRef, bool, error) {
+	source := hr.Spec.GitChartSource
+	if source == nil {
+		return sourceRef{}, false, nil
+	}
+
+	if status, err := repo.Status(); status != git.RepoReady {
+		return sourceRef{}, false, ChartNotReadyError{err}
+	}
+
+	var changed bool
+	s, ok := c.get(hr)
+	if !ok {
+		s = sourceRef{mirror: mirrorName, remote: source.GitURL, ref: source.RefOrDefault()}
+		changed = true
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	head, err := repo.Revision(ctx, s.ref)
+	cancel()
+	if err != nil {
+		return sourceRef{}, false, ChartUnavailableError{err}
+	}
+
+	if !changed {
+		// If the head still equals to what is in our books, there are no changes.
+		if s.head == head {
+			return s, false, nil
+		}
+
+		// Check if the mirror has seen commits in paths we are interested in for
+		// this release.
+		ctx, cancel = context.WithTimeout(context.Background(), c.config.GitTimeout)
+		commits, err := repo.CommitsBetween(ctx, s.head, head, source.Path)
+		cancel()
+		if err != nil {
+			return sourceRef{}, false, ChartUnavailableError{err}
+		}
+		changed = len(commits) > 0
+	}
+
+	s.head = head
+	c.store(hr, s)
+	return s, changed, nil
+}
+
+// maybeMirror requests a new mirror for the given remote. The return value
+// indicates whether the repo was already present (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) maybeMirror(mirrorName string, remote string) bool {
+	ok := c.mirrors.Mirror(
```

Comment:
Maybe this isn't a new problem but, I only see a `StopOne()` counterpart in `Delete()` which seems to only be invoked when a `HelmRelease()` is deleted.

Doesn't this mean that there will be a mirror leak if `git` entry of a `HelmRelease` is changed?

Reply:
Actually, I don't think `Delete()` helps once the `git` field is modified, since it uses `helmReleasesForMirror()` which relies in live data, and not the mirrors currently being used.

Reply:
If we don't want to track changes on the`git`-field explicitly, I think this can be fixed by periodically comparing the output of `helmReleasesForMirror()` on `c.mirrors`. I don't think it's ideal though.

Reply:
Now I see that `forHelmRelease()` is used in `get()` to take into account that  the repo of `HelmRelease`s  can change with respect to what's book-kept in `GitChartSync`, but I don't think we do anything to correct divergences.

Reply:
> Maybe this isn't a new problem but, I only see a `StopOne()` counterpart to this `Mirror()` call in Delete() which seems to only be invoked when a `HelmRelease()` is deleted.

The current version does not stop any mirrors, they continue to live forever. I will let your comments simmer in my head for a while.